[0.11.0] panic: interface conversion: influxql.Expr is *influxql.BinaryExpr, not *influxql.VarRef

Build: 0.11.0~n1457337636

Seeing panic when running this (malformed) query. 
mean(field - field) alone yields: `ERR: error parsing query: expected field argument in mean()`, but nested in non_negative_derivative() causes a panic.

We've just opened up our instance to some more internal developers ;)

```
SELECT non_negative_derivative(mean(created_tmp_tables - created_tmp_disk_tables), 1m) FROM tg_udp."default".mysql WHERE host = 'mysql02' AND time > 1457443029s AND time < 1457609873s GROUP BY time(1m)
```

**Attempt 1:**

```
[query] 2016/03/10 05:39:40 SELECT non_negative_derivative(mean(created_tmp_tables - created_tmp_disk_tables), 1m) FROM tg_udp."default".mysql WHERE host = 'mysql02' AND time > 1457443029s AND time < 1457609873s GROUP BY time(1m)
panic: interface conversion: influxql.Expr is *influxql.BinaryExpr, not *influxql.VarRef
goroutine 3544214 [running]:
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*Engine).CreateIterator(0xc288589810, 0x7fac58fa8250, 0xc277e32270, 0x10ae888, 0x0, 0x0, 0xc345c516e0, 0x1, 0x1, 0xdf8475800, ...)
        /tmp/influxdb/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:677 +0xfe
github.com/influxdata/influxdb/tsdb.(*Shard).CreateIterator(0xc20808de00, 0x7fac58fa8250, 0xc277e32270, 0x10ae888, 0x0, 0x0, 0xc345c516e0, 0x1, 0x1, 0xdf8475800, ...)
        /tmp/influxdb/src/github.com/influxdata/influxdb/tsdb/shard.go:401 +0x12d
github.com/influxdata/influxdb/tsdb.(*shardIteratorCreator).CreateIterator(0xc2499418c8, 0x7fac58fa8250, 0xc277e32270, 0x10ae888, 0x0, 0x0, 0xc345c516e0, 0x1, 0x1, 0xdf8475800, ...)
        /tmp/influxdb/src/github.com/influxdata/influxdb/tsdb/shard.go:810 +0x78
github.com/influxdata/influxdb/influxql.func·046(0x0, 0x0)
        /tmp/influxdb/src/github.com/influxdata/influxdb/influxql/iterator.go:469 +0x100
github.com/influxdata/influxdb/influxql.IteratorCreators.CreateIterator(0xc345c51720, 0x1, 0x1, 0x7fac58fa8250, 0xc277e32270, 0x10ae888, 0x0, 0x0, 0xc345c516e0, 0x1, ...)
        /tmp/influxdb/src/github.com/influxdata/influxdb/influxql/iterator.go:476 +0x110
github.com/influxdata/influxdb/influxql.(*IteratorCreators).CreateIterator(0xc230061e40, 0x7fac58fa8250, 0xc277e32270, 0x10ae888, 0x0, 0x0, 0xc345c516e0, 0x1, 0x1, 0xdf8475800, ...)
        <autogenerated>:134 +0xfd
github.com/influxdata/influxdb/influxql.func·056(0x0, 0x0, 0x0, 0x0)
        /tmp/influxdb/src/github.com/influxdata/influxdb/influxql/select.go:248 +0xc48
github.com/influxdata/influxdb/influxql.buildExprIterator(0x7fac58fa8250, 0xc277e32270, 0x7fac58fa9eb0, 0xc230061e40, 0x7fac58fa8250, 0xc277e32270, 0x10ae888, 0x0, 0x0, 0xc345c516e0, ...)
        /tmp/influxdb/src/github.com/influxdata/influxdb/influxql/select.go:326 +0x1220
github.com/influxdata/influxdb/influxql.buildExprIterator(0x7fac58fa8250, 0xc277e322a0, 0x7fac58fa9eb0, 0xc230061e40, 0x7fac58fa8250, 0xc277e322a0, 0x10ae888, 0x0, 0x0, 0xc345c516e0, ...)
        /tmp/influxdb/src/github.com/influxdata/influxdb/influxql/select.go:220 +0xd16
github.com/influxdata/influxdb/influxql.func·055(0x0, 0x0)
        /tmp/influxdb/src/github.com/influxdata/influxdb/influxql/select.go:151 +0x64c
github.com/influxdata/influxdb/influxql.buildFieldIterators(0xc29a69efb0, 0x1, 0x1, 0x7fac58fa9eb0, 0xc230061e40, 0x0, 0x0, 0x10ae888, 0x0, 0x0, ...)
        /tmp/influxdb/src/github.com/influxdata/influxdb/influxql/select.go:187 +0x103
github.com/influxdata/influxdb/influxql.Select(0xc26fd0cf70, 0x7fac58fa9eb0, 0xc230061e40, 0xc277e32030, 0x0, 0x0, 0x0, 0x0, 0x0)
        /tmp/influxdb/src/github.com/influxdata/influxdb/influxql/select.go:77 +0xd7a
github.com/influxdata/influxdb/cluster.(*QueryExecutor).executeSelectStatement(0xc20805c7e0, 0xc26fd0cf70, 0x2710, 0x1, 0xc305af5ec0, 0xc305af5e60, 0x0, 0x0)
        /tmp/influxdb/src/github.com/influxdata/influxdb/cluster/query_executor.go:447 +0x600
github.com/influxdata/influxdb/cluster.(*QueryExecutor).executeQuery(0xc20805c7e0, 0xc239954380, 0xc28d4c568e, 0x6, 0x2710, 0xc305af5e60, 0xc305af5ec0)
        /tmp/influxdb/src/github.com/influxdata/influxdb/cluster/query_executor.go:120 +0x6d4
created by github.com/influxdata/influxdb/cluster.(*QueryExecutor).ExecuteQuery
        /tmp/influxdb/src/github.com/influxdata/influxdb/cluster/query_executor.go:73 +0x8c
```

**Attempt 2:**

```
[query] 2016/03/10 06:06:14 SELECT non_negative_derivative(mean(created_tmp_tables - created_tmp_disk_tables), 1m) FROM tg_udp."default".mysql WHERE host = 'mysql02' AND time > 1457443029s AND time < 1457609873s GROUP BY time(1m)
panic: interface conversion: influxql.Expr is *influxql.BinaryExpr, not *influxql.VarRef

goroutine 2684 [running]:
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*Engine).CreateIterator(0xc2c8218e70, 0x7fe04f49d040, 0xc23ee78540, 0x10ae888, 0x0, 0x0, 0xc2e47e1950, 0x1, 0x1, 0xdf8475800, ...)
        /tmp/influxdb/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:677 +0xfe
github.com/influxdata/influxdb/tsdb.(*Shard).CreateIterator(0xc208061e00, 0x7fe04f49d040, 0xc23ee78540, 0x10ae888, 0x0, 0x0, 0xc2e47e1950, 0x1, 0x1, 0xdf8475800, ...)
        /tmp/influxdb/src/github.com/influxdata/influxdb/tsdb/shard.go:401 +0x12d
github.com/influxdata/influxdb/tsdb.(*shardIteratorCreator).CreateIterator(0xc266281738, 0x7fe04f49d040, 0xc23ee78540, 0x10ae888, 0x0, 0x0, 0xc2e47e1950, 0x1, 0x1, 0xdf8475800, ...)
        /tmp/influxdb/src/github.com/influxdata/influxdb/tsdb/shard.go:810 +0x78
github.com/influxdata/influxdb/influxql.func·046(0x0, 0x0)
        /tmp/influxdb/src/github.com/influxdata/influxdb/influxql/iterator.go:469 +0x100
github.com/influxdata/influxdb/influxql.IteratorCreators.CreateIterator(0xc2e47e1970, 0x1, 0x1, 0x7fe04f49d040, 0xc23ee78540, 0x10ae888, 0x0, 0x0, 0xc2e47e1950, 0x1, ...)
        /tmp/influxdb/src/github.com/influxdata/influxdb/influxql/iterator.go:476 +0x110
github.com/influxdata/influxdb/influxql.(*IteratorCreators).CreateIterator(0xc23a007ea0, 0x7fe04f49d040, 0xc23ee78540, 0x10ae888, 0x0, 0x0, 0xc2e47e1950, 0x1, 0x1, 0xdf8475800, ...)
        <autogenerated>:134 +0xfd
github.com/influxdata/influxdb/influxql.func·056(0x0, 0x0, 0x0, 0x0)
        /tmp/influxdb/src/github.com/influxdata/influxdb/influxql/select.go:248 +0xc48
github.com/influxdata/influxdb/influxql.buildExprIterator(0x7fe04f49d040, 0xc23ee78540, 0x7fe04f49e580, 0xc23a007ea0, 0x7fe04f49d040, 0xc23ee78540, 0x10ae888, 0x0, 0x0, 0xc2e47e1950, ...)
        /tmp/influxdb/src/github.com/influxdata/influxdb/influxql/select.go:326 +0x1220
github.com/influxdata/influxdb/influxql.buildExprIterator(0x7fe04f49d040, 0xc23ee78570, 0x7fe04f49e580, 0xc23a007ea0, 0x7fe04f49d040, 0xc23ee78570, 0x10ae888, 0x0, 0x0, 0xc2e47e1950, ...)
        /tmp/influxdb/src/github.com/influxdata/influxdb/influxql/select.go:220 +0xd16
github.com/influxdata/influxdb/influxql.func·055(0x0, 0x0)
        /tmp/influxdb/src/github.com/influxdata/influxdb/influxql/select.go:151 +0x64c
github.com/influxdata/influxdb/influxql.buildFieldIterators(0xc21e054898, 0x1, 0x1, 0x7fe04f49e580, 0xc23a007ea0, 0x0, 0x0, 0x10ae888, 0x0, 0x0, ...)
        /tmp/influxdb/src/github.com/influxdata/influxdb/influxql/select.go:187 +0x103
github.com/influxdata/influxdb/influxql.Select(0xc24500c4e0, 0x7fe04f49e580, 0xc23a007ea0, 0xc23ee78300, 0x0, 0x0, 0x0, 0x0, 0x0)
        /tmp/influxdb/src/github.com/influxdata/influxdb/influxql/select.go:77 +0xd7a
github.com/influxdata/influxdb/cluster.(*QueryExecutor).executeSelectStatement(0xc20806c7e0, 0xc24500c4e0, 0x2710, 0x0, 0xc2378b8ae0, 0xc2378b8a80, 0x0, 0x0)
        /tmp/influxdb/src/github.com/influxdata/influxdb/cluster/query_executor.go:447 +0x600
github.com/influxdata/influxdb/cluster.(*QueryExecutor).executeQuery(0xc20806c7e0, 0xc21cdb1900, 0xc22ff937ab, 0x0, 0x2710, 0xc2378b8a80, 0xc2378b8ae0)
        /tmp/influxdb/src/github.com/influxdata/influxdb/cluster/query_executor.go:120 +0x6d4
created by github.com/influxdata/influxdb/cluster.(*QueryExecutor).ExecuteQuery
        /tmp/influxdb/src/github.com/influxdata/influxdb/cluster/query_executor.go:73 +0x8c
```

--- a/influxql/ast.go
+++ b/influxql/ast.go
@@ -1420,11 +1420,12 @@ func (s *SelectStatement) validateAggregates(tr targetRequirement) error {
 				if err != nil {
 					return fmt.Errorf("invalid group interval: %v", err)
 				}
-				if groupByInterval > 0 {
-					c, ok := expr.Args[0].(*Call)
-					if !ok {
-						return fmt.Errorf("aggregate function required inside the call to %s", expr.Name)
-					}
+
+				if c, ok := expr.Args[0].(*Call); ok && groupByInterval == 0 {
+					return fmt.Errorf("%s aggregate requires a GROUP BY interval", expr.Name)
+				} else if !ok && groupByInterval > 0 {
+					return fmt.Errorf("aggregate function required inside the call to %s", expr.Name)
+				} else if ok {
 					switch c.Name {
 					case "top", "bottom":
 						if err := s.validTopBottomAggr(c); err != nil {
@@ -1437,6 +1438,21 @@ func (s *SelectStatement) validateAggregates(tr targetRequirement) error {
 					default:
 						if exp, got := 1, len(c.Args); got != exp {
 							return fmt.Errorf("invalid number of arguments for %s, expected %d, got %d", c.Name, exp, got)
+						}
+
+						switch fc := c.Args[0].(type) {
+						case *VarRef:
+							// do nothing
+						case *Call:
+							if fc.Name != "distinct" {
+								return fmt.Errorf("expected field argument in %s()", c.Name)
+							}
+						case *Distinct:
+							if expr.Name != "count" {
+								return fmt.Errorf("expected field argument in %s()", c.Name)
+							}
+						default:
+							return fmt.Errorf("expected field argument in %s()", c.Name)
 						}
 					}
 				}

--- a/influxql/parser_test.go
+++ b/influxql/parser_test.go
@@ -190,28 +190,6 @@ func TestParser_ParseStatement(t *testing.T) {
 			},
 		},
 
-		{
-			s: `SELECT derivative(mean(field1), 1h) FROM myseries;`,
-			stmt: &influxql.SelectStatement{
-				IsRawQuery: false,
-				Fields: []*influxql.Field{
-					{Expr: &influxql.Call{Name: "derivative", Args: []influxql.Expr{&influxql.Call{Name: "mean", Args: []influxql.Expr{&influxql.VarRef{Val: "field1"}}}, &influxql.DurationLiteral{Val: time.Hour}}}},
-				},
-				Sources: []influxql.Source{&influxql.Measurement{Name: "myseries"}},
-			},
-		},
-
-		{
-			s: `SELECT derivative(mean(field1)) FROM myseries;`,
-			stmt: &influxql.SelectStatement{
-				IsRawQuery: false,
-				Fields: []*influxql.Field{
-					{Expr: &influxql.Call{Name: "derivative", Args: []influxql.Expr{&influxql.Call{Name: "mean", Args: []influxql.Expr{&influxql.VarRef{Val: "field1"}}}}}},
-				},
-				Sources: []influxql.Source{&influxql.Measurement{Name: "myseries"}},
-			},
-		},
-
 		// SELECT statement (lowercase)
 		{
 			s: `select my_field from myseries`,
@@ -1721,6 +1699,7 @@ func TestParser_ParseStatement(t *testing.T) {
 		{s: `SELECT derivative(bottom(value)) FROM myseries where time < now() and time > now() - 1d group by time(1h)`, err: `invalid number of arguments for bottom, expected at least 2, got 1`},
 		{s: `SELECT derivative(max()) FROM myseries where time < now() and time > now() - 1d group by time(1h)`, err: `invalid number of arguments for max, expected 1, got 0`},
 		{s: `SELECT derivative(percentile(value)) FROM myseries where time < now() and time > now() - 1d group by time(1h)`, err: `invalid number of arguments for percentile, expected 2, got 1`},
+		{s: `SELECT derivative(mean(value), 1h) FROM myseries where time < now() and time > now() - 1d`, err: `derivative aggregate requires a GROUP BY interval`},
 		{s: `SELECT non_negative_derivative(), field1 FROM myseries`, err: `mixing aggregate and non-aggregate queries is not supported`},
 		{s: `select non_negative_derivative() from myseries`, err: `invalid number of arguments for non_negative_derivative, expected at least 1 but no more than 2, got 0`},
 		{s: `select non_negative_derivative(mean(value), 1h, 3) from myseries`, err: `invalid number of arguments for non_negative_derivative, expected at least 1 but no more than 2, got 3`},
@@ -1728,11 +1707,15 @@ func TestParser_ParseStatement(t *testing.T) {
 		{s: `SELECT non_negative_derivative(top(value)) FROM myseries where time < now() and time > now() - 1d group by time(1h)`, err: `invalid number of arguments for top, expected at least 2, got 1`},
 		{s: `SELECT non_negative_derivative(bottom(value)) FROM myseries where time < now() and time > now() - 1d group by time(1h)`, err: `invalid number of arguments for bottom, expected at least 2, got 1`},
 		{s: `SELECT non_negative_derivative(max()) FROM myseries where time < now() and time > now() - 1d group by time(1h)`, err: `invalid number of arguments for max, expected 1, got 0`},
+		{s: `SELECT non_negative_derivative(mean(value), 1h) FROM myseries where time < now() and time > now() - 1d`, err: `non_negative_derivative aggregate requires a GROUP BY interval`},
 		{s: `SELECT non_negative_derivative(percentile(value)) FROM myseries where time < now() and time > now() - 1d group by time(1h)`, err: `invalid number of arguments for percentile, expected 2, got 1`},
 		{s: `SELECT field1 from myseries WHERE host =~ 'asd' LIMIT 1`, err: `found asd, expected regex at line 1, char 42`},
 		{s: `SELECT value > 2 FROM cpu`, err: `invalid operator > in SELECT clause at line 1, char 8; operator is intended for WHERE clause`},
 		{s: `SELECT value = 2 FROM cpu`, err: `invalid operator = in SELECT clause at line 1, char 8; operator is intended for WHERE clause`},
 		{s: `SELECT s =~ /foo/ FROM cpu`, err: `invalid operator =~ in SELECT clause at line 1, char 8; operator is intended for WHERE clause`},
+		// TODO: Remove this restriction in the future: https://github.com/influxdata/influxdb/issues/5968
+		{s: `SELECT mean(cpu_total - cpu_idle) FROM cpu`, err: `expected field argument in mean()`},
+		{s: `SELECT derivative(mean(cpu_total - cpu_idle), 1s) FROM cpu WHERE time < now() AND time > now() - 1d GROUP BY time(1h)`, err: `expected field argument in mean()`},
 		// See issues https://github.com/influxdata/influxdb/issues/1647
 		// and https://github.com/influxdata/influxdb/issues/4404
 		//{s: `DELETE`, err: `found EOF, expected FROM at line 1, char 8`},

